internal/build/imgsrc/depot.go

```diff
@@ -0,0 +1,331 @@
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+var _ imageBuilder = (*DepotBuilder)(nil)
+
+type DepotBuilder struct{}
+
+func (d *DepotBuilder) Name() string { return "depot.dev" }
+
+func (d *DepotBuilder) Run(ctx context.Context, _ *dockerClientFactory, streams *iostreams.IOStreams, opts ImageOptions, build *build) (*DeploymentImage, string, error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "depot_builder", trace.WithAttributes(opts.ToSpanAttributes()...))
+	defer span.End()
+
+	build.BuildStart()
+
+	var dockerfile string
+
+	if opts.DockerfilePath != "" {
+		if !helpers.FileExists(opts.DockerfilePath) {
+			build.BuildFinish()
+			err := fmt.Errorf("dockerfile '%s' not found", opts.DockerfilePath)
+			tracing.RecordError(span, err, "failed to find dockerfile")
+			return nil, "", err
+		}
+		dockerfile = opts.DockerfilePath
+	} else {
+		dockerfile = ResolveDockerfile(opts.WorkingDir)
+	}
+
+	if dockerfile == "" {
+		span.AddEvent("dockerfile not found, skipping")
+		terminal.Debug("dockerfile not found, skipping")
+		build.BuildFinish()
+		return nil, "", nil
+	}
+
+	var relDockerfile string
+	if isPathInRoot(dockerfile, opts.WorkingDir) {
+		// pass the relative path to Dockerfile within the context
+		p, err := filepath.Rel(opts.WorkingDir, dockerfile)
+		if err != nil {
+			tracing.RecordError(span, err, "failed to get relative dockerfile path")
+			build.BuildFinish()
+			return nil, "", err
+		}
+		// On Windows, convert \ to a slash / as the docker build will
+		// run in a Linux VM at the end.
+		relDockerfile = filepath.ToSlash(p)
+	}
+	span.SetAttributes(attribute.String("relative_dockerfile_path", relDockerfile))
+
+	build.ImageBuildStart()
+
+	image, err := depotBuild(ctx, streams, opts, dockerfile, build)
+	if err != nil {
+		metrics.SendNoData(ctx, "remote_builder_failure")
+		build.ImageBuildFinish()
+		build.BuildFinish()
+		tracing.RecordError(span, err, "failed to build image")
+		return nil, "", errors.Wrap(err, "error building")
+	}
+
+	build.ImageBuildFinish()
+	build.BuildFinish()
+	cmdfmt.PrintDone(streams.ErrOut, "Building image done")
+
+	span.SetAttributes(image.ToSpanAttributes()...)
+	return image, "", nil
+}
+
+func depotBuild(ctx context.Context, streams *iostreams.IOStreams, opts ImageOptions, dockerfilePath string, buildState *build) (*DeploymentImage, error) {
+	buildState.BuilderInitStart()
+
+	{
+		msg := "Waiting for depot builder...\n"
+		if streams.IsInteractive() {
+			streams.StartProgressIndicatorMsg(msg)
+		} else {
+			fmt.Fprintln(streams.ErrOut, msg)
+		}
+	}
+
+	apiClient := flyutil.ClientFromContext(ctx)
+	region := os.Getenv("FLY_REMOTE_BUILDER_REGION")
+	if region != "" {
+		region = "fly-" + region
+	}
+
+	buildInfo, err := apiClient.EnsureDepotRemoteBuilder(ctx, &fly.EnsureDepotRemoteBuilderInput{
+		AppName: &opts.AppName,
+		Region:  &region,
+	})
+	if err != nil {
+		streams.StopProgressIndicator()
+		buildState.BuilderInitFinish()
```

Comment:
Could `buildStateInitFinish()` be in a defer statement?

Reply:
that's a good idea. i'll move the builder init code into a new function